Shared/sdk/SharedUtil.Template.h

```diff
@@ -0,0 +1,193 @@
+#pragma once
+
+/**
+    is_Nspecialization
+
+    These structs allow testing whether a type is a specialization of a
+    class template
+
+    Note: Take care of optional parameters. For example std::unordered_map
+    has 5 template arguments, thus is_5specialization needs to be used, rather
+    than is_2specialization (Key, Value).
+**/
+
+template <typename Test, template <typename> class Ref>
+struct is_specialization : std::false_type
+{
+};
+
+template <template <typename> class Ref, typename Args>
+struct is_specialization<Ref<Args>, Ref> : std::true_type
+{
+    using param_t = Args;
+};
+
+template <typename Test, template <typename, typename> class Ref>
+struct is_2specialization : std::false_type
+{
+};
+
+template <template <typename, typename> class Ref, typename Arg1, typename Arg2>
+struct is_2specialization<Ref<Arg1, Arg2>, Ref> : std::true_type
+{
+    using param1_t = Arg1;
+    using param2_t = Arg2;
+};
+
+template <typename Test, template <typename, typename, typename, typename, typename> class Ref>
+struct is_5specialization : std::false_type
+{
+};
+
+template <template <typename, typename, typename, typename, typename> class Ref, typename Arg1, typename Arg2, typename Arg3, typename Arg4, typename Arg5>
+struct is_5specialization<Ref<Arg1, Arg2, Arg3, Arg4, Arg5>, Ref> : std::true_type
+{
+    using param1_t = Arg1;
+    using param2_t = Arg2;
+};
+
+template <typename Test>
+struct is_variant : std::false_type
+{
+};
+
+template <typename Arg1, typename... Args>
+struct is_variant<std::variant<Arg1, Args...>> : std::true_type
+{
+    using param1_t = Arg1;
+    using rest_t = std::variant<Args...>;
+    static constexpr auto count = sizeof...(Args) + 1;
+};
+
+/**
+    nth_element
+
+    Returns the nth element of a parameter pack
```

Comment:
This is also another very cool use of recursion.

Consider making this obvious by:
- describing the implementation briefly in the main comment
- moving the base case _above_ the general case
- commenting both implementations with either `// Base case (I == 0)` or `// General case (I > 0)`

```suggestion
    Returns the nth element of a parameter pack.

    It does this by recursively selecting the `n - 1`th item.
    
    For example, to get the Nth item of parameter pack `Args` you:
        nth_element_t<N, Args...>>
```